lxd/apparmor/archive.go

```diff
@@ -0,0 +1,108 @@
+package apparmor
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"text/template"
+
+	"github.com/lxc/lxd/lxd/state"
+)
+
+var archiveProfileTpl = template.Must(template.New("archiveProfile").Parse(`#include <tunables/global>
+profile "{{.name}}" {
+  /usr/lib/x86_64-linux-gnu/** mr,
+  /etc/ld.so.cache r,
+  /usr/lib/locale/locale-archive r,
+  /etc/locale.alias r,
+  @{PROC}/filesystems r,
+  @{PROC}/@{pid}/mounts r,
+  /etc/nsswitch.conf r,
+  /etc/passwd r,
+  /etc/group r,
+
+  /usr/bin/gzip ux,
+  /usr/bin/bzip2 ux,
+  /usr/bin/xz ux,
+  /usr/bin/zstd ux,
+
+  unix (create),
+
+  signal (receive, send) peer=@{profile_name},
+
+  {{.outputPath}}/ rw,
+  {{.outputPath}}/** rw,
+
+  capability dac_read_search,
+  capability dac_override,
+  capability chown,
+  capability fsetid,
+}
+`))
+
+// ArchiveLoad ensures that the archive's policy is loaded into the kernel.
+func ArchiveLoad(state *state.State, name string) error {
+	profile := filepath.Join(aaPath, "profiles", ArchiveProfileFilename(name))
+	content, err := ioutil.ReadFile(profile)
+	if err != nil && !os.IsNotExist(err) {
+		return err
+	}
+
+	updated, err := archiveProfile(name)
+	if err != nil {
+		return err
+	}
+
+	if string(content) != string(updated) {
+		err = ioutil.WriteFile(profile, []byte(updated), 0600)
+		if err != nil {
+			return err
+		}
+	}
+
+	err = loadProfile(state, ArchiveProfileFilename(name))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ArchiveUnload ensures that the archive's policy namespace is unloaded to free kernel memory.
+// This does not delete the policy from disk or cache.
+func ArchiveUnload(state *state.State, name string) error {
+	err := unloadProfile(state, ArchiveProfileName(name), ArchiveProfileFilename(name))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// archiveProfile generates the AppArmor profile template from the given destination path.
+func archiveProfile(name string) (string, error) {
```

Comment:
arg name should be `outputPath` to indicate it expects a path that represents the output of the compress archive.

Reply:
Fixed.